src/Base/PyHandle.h

```diff
@@ -0,0 +1,152 @@
+ * against the latter).
+ *
+ * Usage example:
+ *     //create a py object, and save it into the handle
+ *     UnsafePyHandle<Base::Placement> h (new Base::PlacementPy(new Base::Placement), true);
+ *     //use the handle as if it is a pointer to C++ object
+ *     h->getPosition()
+ *     //and when the handle is destroyed, so is the py object, if it isn't referred to by something else.
+ *
+ * ## Memory management
+ *
+ * * when PyHandle is created, reference counter is incremented (except if
+ * "new_reference" constructor argument is true).
+ *
+ * * when PyHandle is destroyed, reference counter is decremented (no matter
+ * what "new_reference" value was passed to constructor). If resulting refcount
+ * is zero, the object is deleted.
+ *
+ * * Py::new_reference_to(PyHandle) increments reference count and returns PyObject*
+ *
+ * Cheatsheet:
+ *
+ * * when constructing from "new SomethingSomethingPy(new SomethingSomething)",
+ * use "new_reference". That is because SomethingSomethingPy sets reference counter
+ * to 1 in its constructor.
+ *
+ * * when constructing from getPyObject, use "new_reference"
+ *
+ * * when constructing from PyObject* you got as an argument of ...PyImp method
+ * implementation, don't use "new_reference".
+ *
+ * * when returning PyObject* from a getPyObject, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * * when returning PyObject* from a ...PyImp method implementation, use
+ * Py::new_reference_to(<PyHangle> object).
+ *
+ * The handle derives from PyCXX's Py::Object, so it can be conveniently used
+ * for adding to tuples and the like, when programming python API.
+ */
+ // Since it was made to become the main memory management construct of
+ // ConstraintSolver, checks for isValid() are intentionally not done to maximize
+ // performance. It is named "Unsafe" with the intention to have room for a safer
+ // "PyHandle" replica.
+template <class CppType>
+class UnsafePyHandle : public Py::Object
+{
+private:
+    CppType* cppptr;
+protected:
+    virtual void setObject(PyObject* pyob, bool new_reference = false){
+        if (typecheck(pyob))
+            cppptr = static_cast<CppType*>(static_cast<Base::PyObjectBase*>(pyob)->twinPtr());
+        else
+            throw Base::TypeError("Not a freecad object");
+        this->set(pyob, new_reference);
+    }
+
+public:
+    virtual bool typecheck(PyObject* pyob) {
+        if (! PyObject_TypeCheck(pyob, &(Base::PyObjectBase::Type)))
+            return false;
+        if (! static_cast<Base::PyObjectBase*>(pyob)->isValid())
+            return false;
+        return true;
+    }
+
+    //We can't override Py::Object::set
+    //so, mask out all methods to change pointer of Py::Object
+    //it may still be possible to bypass with typecasts,
+    //but at least that isn't straightforward.
+    /**
+     * @brief UnsafePyHandle constructor
+     * @param pyob
+     * @param new_reference: specify True if the reference is "owned" by the calling function,
```

Comment:
you don't need the `:` after `new_reference`, doxygen will take care of the formatting for you.

Reply:
ok, thanks for your doxygen help. I never tried reading the doxygen output, and don't really know the syntax. So I mostly make it look good in the source code.